autovacuum_message_attachment/models/vacuum_rule.py

```diff
@@ -0,0 +1,134 @@
+             "keeped in the database after creation. Once the delay is "
+             "passed, they will be automatically deleted.")
+    active = fields.Boolean(default=True)
+    description = fields.Text()
+
+    @api.multi
+    @api.constrains('retention_time')
+    def retention_time_not_null(self):
+        for rule in self:
+            if not rule.retention_time:
+                raise exceptions.ValidationError(
+                    _("The Retention Time can't be 0 days"))
+
+    @api.multi
+    def _get_message_domain(self):
+        today = date.today()
+        limit_date = (today - timedelta(days=self.retention_time)).strftime(
+            DEFAULT_SERVER_DATE_FORMAT)
+        message_domain = [('date', '<', limit_date)]
+        if self.message_type != 'all':
+            message_domain += [('message_type', '=', self.message_type)]
+        if self.model_ids:
+            models = self.model_ids.mapped('model')
+            message_domain += [('model', 'in', models)]
+
+        subtype_ids = self.message_subtype_ids.ids
+        if subtype_ids and self.empty_subtype:
+            message_domain = ['|', ('subtype_id', 'in', subtype_ids),
+                              ('subtype_id', '=', False)]
+        elif subtype_ids and not self.empty_subtype:
+            message_domain += [('subtype_id', 'in', subtype_ids)]
+        elif not subtype_ids and not self.empty_subtype:
+            message_domain += [('subtype_id', '!=', False)]
+        return message_domain
+
+    @api.multi
+    def _get_attachment_domain(self):
+        today = date.today()
+        limit_date = (today - timedelta(days=self.retention_time)).strftime(
+            DEFAULT_SERVER_DATE_FORMAT)
+        attachment_domain = [('create_date', '<', limit_date)]
+        if self.filename_pattern:
+            attachment_domain += [('name', 'ilike', self.filename_pattern)]
+        if self.model_ids:
+            models = self.model_ids.mapped('model')
+            attachment_domain += [('res_model', 'in', models)]
+        else:
+            # Avoid deleting attachment without model, if there are, it is
+            # probably some attachments created by Odoo
+            attachment_domain += [('res_model', '!=', False)]
+        return attachment_domain
+
+    @api.multi
+    def get_domain(self):
+        self.ensure_one()
+        domain = []
+        if self.ttype == 'message':
+            domain += self._get_message_domain()
+        elif self.ttype == 'attachment':
+            domain += self._get_attachment_domain()
+
+        # Case we want a condition on linked model records
+        if self.model_id and self.model_filter_domain:
+            record_domain = safe_eval(self.model_filter_domain,
+                                      locals_dict={'datetime': datetime})
+
+            res_ids = self.env[self.model_id.model].with_context(
+                active_test=False).search(record_domain).ids
+            domain += ['|', ('res_id', 'in', res_ids),
```

Comment:
@etobella 
Actually, before installing this module, you shoud cleanup yourself the messages/attachments according to your rules. Because anyway, the deletion will take for ever and the cron won't ever pass...

For example, if you have 30 millions mail message over 5 years and you decide to keep one year history and install the module, the cron won't be able to delete 24 millions mail message...
In this cases, you should delete your 4 years history as you wish (a script with direct sql queries seems more appropriates, maybe adding some indexes for the occasion...).
And even this way, it will take hours (or days)

And, then, install and parameter the module, so it deletes one day history...each night!

I've done this way for some customers and I have no performance issue AFAIK (with the cron running at night)

I should probably say this in the README. 

Anyway, I am not against another proposal, just need to keep the feature.

Reply:
I know, but that's not the problem, Assume you want to clean messages only on sale.orders that are cancelled, then your system will get all the sale.orders cancelled and add this on the query. Every night, you will check all the cancelled orders. Assume you have a million cancelled orders, then imagine how the query will be...

Reply:
Yes, that's right...
I did kind of get around the issue, but I know it is not ideal.
For one customer, I do delete messages for done/canceled pickings, procurements (version 8), etc.. (huge tables) doing something like this : 
`[('state', 'in', ('done', 'cancel')), ('write_date', '<', ((datetime.datetime.now()-datetime.timedelta(days=365)).strftime('%Y-%m-%d'))), ('write_date', '>', ((datetime.datetime.now()-datetime.timedelta(days=375)).strftime('%Y-%m-%d')))]`

Far from perfect, and kind of complicated, but working well.
Once again, if it can be improved, that is better of course.

Reply:
I made a refactoring on my branch:
https://github.com/tegin/server-tools/commits/12-autovacuum_message_attachment

Can you check it please :wink:

Reply:
Thanks for the proposal.
I am not familiar with the class `Base`, where do we see that every other models depend on this one?

Overall it looks good, but I don't really see why it would be faster.
I'd need to do some test on a big database about it. I can backport this to v8 to make tests with old and new version and compare.

Reply:
Ok! Thanks, the main difference is that we only find the records from the model that has messages with issues.
With the case I checked before: We will find all the sale.orders that are cancelled with messages older than a year, then we search only this messages.

Base is parent from all: https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L516

Reply:
@etobella well, I did some tests with a huge table in v8 and it seems really faster indeed.
Can you make a PR in my branch so I can merge it?
Thanks

Reply:
Sure!! Akretion#22